### PR TITLE
ZJIT: Specialize polymorphic sends with block literals

### PR DESCRIPTION
**Pull request by tekknolagi**
Reduce polymorphic sends on lobsters by ~850k. Unfortunately,
`send_block_arg_not_nil` goes up by ~500k. So the net benefit here is
~350k. The `send_block_arg_not_nil` needs investigating.

### CODE REVIEW

**Review by k0kubun — APPROVED** [review]
looks good. This seems to handle https://github.com/ruby/ruby/pull/18057 properly too.

I think we could let `opt_send_without_block` and `send` share some of the logic (like YJIT does), but I think it makes sense to land this first and then discuss how we refactor it separately.

File: zjit/src/hir.rs [hunk 1]
```diff
@@ -10191,8 +10191,55 @@ fn add_iseq_to_hir(
...
+                    if let Some(summary) = fun.polymorphic_summary(&profiles, recv, exit_id) {
+                        let join_block = fun.new_block(insn_idx);
+                        let join_param = fun.push_insn(join_block, Insn::Param);
+                        // Dedup by expected type so immediate/heap variants
+                        // under the same Ruby class can still get separate branches.
+                        let mut seen_types = Vec::with_capacity(summary.buckets().len());
+                        for &profiled_type in summary.buckets() {
+                            if profiled_type.is_empty() { break; }
+                            let expected = Type::from_profiled_type(profiled_type);
+                            if seen_types.iter().any(|ty: &Type| ty.bit_equal(expected)) {
```
Comment (nirvdrum) [inline zjit/src/hir.rs:10203]: I found it funny that `bit_equal` has a comment reading "Do not use!". But, it looks okay here.

**Comment by nirvdrum** [comment]
For anyone interested, this is the HIR effect of the change on the newly added `test_specialize_polymorphic_send_with_block` test:

<details><summary>Old</summary>
<p>

```
fn test@<compiled>:8:
bb1():
  EntryPoint interpreter
  v1:BasicObject = LoadSelf
  v2:CPtr = LoadSP
  v3:BasicObject = LoadField v2, :obj@0x1000
  Jump bb3(v1, v3)
bb2():
  EntryPoint JIT(0)
  v6:BasicObject = LoadArg :self@0
  v7:BasicObject = LoadArg :obj@1
  Jump bb3(v6, v7)
bb3(v9:BasicObject, v10:BasicObject):
  v15:BasicObject = Send v10, 0x1008, :foo # SendFallbackReason: Send: polymorphic call site
  PatchPoint NoEPEscape(test)
  CheckInterrupts
  Return v15
```

</p>
</details> 

<details><summary>New</summary>
<p>

```
fn test@<compiled>:8:
bb1():
  EntryPoint interpreter
  v1:BasicObject = LoadSelf
  v2:CPtr = LoadSP
  v3:BasicObject = LoadField v2, :obj@0x1000
  Jump bb3(v1, v3)
bb2():
  EntryPoint JIT(0)
  v6:BasicObject = LoadArg :self@0
  v7:BasicObject = LoadArg :obj@1
  Jump bb3(v6, v7)
bb3(v9:BasicObject, v10:BasicObject):
  v16:CBool = HasType v10, ObjectSubclass[class_exact:C]
  CondBranch v16, bb5(), bb6()
bb5():
  v19:ObjectSubclass[class_exact:C] = RefineType v10, ObjectSubclass[class_exact:C]
  PatchPoint NoSingletonClass(C@0x1008)
  PatchPoint MethodRedefined(C@0x1008, foo@0x1010, cme:0x1018)
  PushInlineFrame :foo, v19 (0x1040), num_args=0
  v57:CPtr = GetEP 0
  v58:CInt64 = LoadField v57, :VM_ENV_DATA_INDEX_SPECVAL@0x1060
  v59:CInt64[-4] = Const CInt64(-4)
  v60:CInt64 = IntAnd v58, v59
  v61:BasicObject = InvokeBlockIseqDirect (0x1068), v60
  CheckInterrupts
  PopInlineFrame
  Jump bb4(v61)
bb6():
  v22:CBool = HasType v10, ObjectSubclass[class_exact:A]
  CondBranch v22, bb7(), bb8()
bb7():
  v25:ObjectSubclass[class_exact:A] = RefineType v10, ObjectSubclass[class_exact:A]
  PatchPoint NoSingletonClass(A@0x1088)
  PatchPoint MethodRedefined(A@0x1088, foo@0x1010, cme:0x1018)
  PushInlineFrame :foo, v25 (0x1040), num_args=0
  v75:CPtr = GetEP 0
  v76:CInt64 = LoadField v75, :VM_ENV_DATA_INDEX_SPECVAL@0x1060
  v77:CInt64[-4] = Const CInt64(-4)
  v78:CInt64 = IntAnd v76, v77
  v79:BasicObject = InvokeBlockIseqDirect (0x1068), v78
  CheckInterrupts
  PopInlineFrame
  Jump bb4(v79)
bb8():
  v28:CBool = HasType v10, ObjectSubclass[class_exact:B]
  CondBranch v28, bb9(), bb10()
bb9():
  v31:ObjectSubclass[class_exact:B] = RefineType v10, ObjectSubclass[class_exact:B]
  PatchPoint NoSingletonClass(B@0x1090)
  PatchPoint MethodRedefined(B@0x1090, foo@0x1010, cme:0x1018)
  PushInlineFrame :foo, v31 (0x1040), num_args=0
  v93:CPtr = GetEP 0
  v94:CInt64 = LoadField v93, :VM_ENV_DATA_INDEX_SPECVAL@0x1060
  v95:CInt64[-4] = Const CInt64(-4)
  v96:CInt64 = IntAnd v94, v95
  v97:BasicObject = InvokeBlockIseqDirect (0x1068), v96
  CheckInterrupts
  PopInlineFrame
  Jump bb4(v97)
bb10():
  v34:BasicObject = Send v10, 0x1068, :foo # SendFallbackReason: Send: polymorphic call site
  Jump bb4(v34)
bb4(v15:BasicObject):
  PatchPoint NoEPEscape(test)
  CheckInterrupts
  Return v15
```

</p>
</details>

```diff
14c14,64
<           v15:BasicObject = Send v10, 0x1008, :foo # SendFallbackReason: Send: polymorphic call site
---
>           v16:CBool = HasType v10, ObjectSubclass[class_exact:C]
>           CondBranch v16, bb5(), bb6()
>         bb5():
>           v19:ObjectSubclass[class_exact:C] = RefineType v10, ObjectSubclass[class_exact:C]
>           PatchPoint NoSingletonClass(C@0x1008)
>           PatchPoint MethodRedefined(C@0x1008, foo@0x1010, cme:0x1018)
>           PushInlineFrame :foo, v19 (0x1040), num_args=0
>           v57:CPtr = GetEP 0
>           v58:CInt64 = LoadField v57, :VM_ENV_DATA_INDEX_SPECVAL@0x1060
>           v59:CInt64[-4] = Const CInt64(-4)
>           v60:CInt64 = IntAnd v58, v59
>           v61:BasicObject = InvokeBlockIseqDirect (0x1068), v60
>           CheckInterrupts
>           PopInlineFrame
>           Jump bb4(v61)
>         bb6():
>           v22:CBool = HasType v10, ObjectSubclass[class_exact:A]
>           CondBranch v22, bb7(), bb8()
>         bb7():
>           v25:ObjectSubclass[class_exact:A] = RefineType v10, ObjectSubclass[class_exact:A]
>           PatchPoint NoSingletonClass(A@0x1088)
>           PatchPoint MethodRedefined(A@0x1088, foo@0x1010, cme:0x1018)
>           PushInlineFrame :foo, v25 (0x1040), num_args=0
>           v75:CPtr = GetEP 0
>           v76:CInt64 = LoadField v75, :VM_ENV_DATA_INDEX_SPECVAL@0x1060
>           v77:CInt64[-4] = Const CInt64(-4)
>           v78:CInt64 = IntAnd v76, v77
>           v79:BasicObject = InvokeBlockIseqDirect (0x1068), v78
>           CheckInterrupts
>           PopInlineFrame
>           Jump bb4(v79)
>         bb8():
>           v28:CBool = HasType v10, ObjectSubclass[class_exact:B]
>           CondBranch v28, bb9(), bb10()
>         bb9():
>           v31:ObjectSubclass[class_exact:B] = RefineType v10, ObjectSubclass[class_exact:B]
>           PatchPoint NoSingletonClass(B@0x1090)
>           PatchPoint MethodRedefined(B@0x1090, foo@0x1010, cme:0x1018)
>           PushInlineFrame :foo, v31 (0x1040), num_args=0
>           v93:CPtr = GetEP 0
>           v94:CInt64 = LoadField v93, :VM_ENV_DATA_INDEX_SPECVAL@0x1060
>           v95:CInt64[-4] = Const CInt64(-4)
>           v96:CInt64 = IntAnd v94, v95
>           v97:BasicObject = InvokeBlockIseqDirect (0x1068), v96
>           CheckInterrupts
>           PopInlineFrame
>           Jump bb4(v97)
>         bb10():
>           v34:BasicObject = Send v10, 0x1068, :foo # SendFallbackReason: Send: polymorphic call site
>           Jump bb4(v34)
>         bb4(v15:BasicObject):

```

File: zjit/src/hir/opt_tests.rs [hunk 2]
```diff
@@ -22855,4 +22855,91 @@ mod hir_opt_tests {
...
+          v93:CPtr = GetEP 0
+          v94:CInt64 = LoadField v93, :VM_ENV_DATA_INDEX_SPECVAL@0x1060
+          v95:CInt64[-4] = Const CInt64(-4)
+          v96:CInt64 = IntAnd v94, v95
+          v97:BasicObject = InvokeBlockIseqDirect (0x1068), v96
+          CheckInterrupts
+          PopInlineFrame
+          Jump bb4(v97)
+        bb10():
+          v34:BasicObject = Send v10, 0x1068, :foo # SendFallbackReason: Send: polymorphic call site
```
Comment (nirvdrum) [inline zjit/src/hir/opt_tests.rs:22937]: I'm not sure the fallback reason set on hir.rs:10233 is surviving. I think it's supposed to be `SendFallbackReason: Send: polymorphic fallback`. I guess it doesn't matter a whole lot, but it could skew attribution or maybe hints at a deeper issue.

File: zjit/src/hir/opt_tests.rs [hunk 3]
```diff
@@ -22855,4 +22855,91 @@ mod hir_opt_tests {
           Return v37
         ");
     }
+
+    #[test]
+    fn test_specialize_polymorphic_send_with_block() {
```
Comment (nirvdrum) [inline zjit/src/hir/opt_tests.rs:22860]: I think it's worth adding a test that passes a block by arg, both `nil` and non-`nil`.

File: zjit/src/hir.rs (hunk 1 above)
Comment (tekknolagi) [reply zjit/src/hir.rs:10203]: Yeah the majority of uses for checking type-related things (vs de-duplicating) should be using subtyping

File: zjit/src/hir/opt_tests.rs (hunk 2 above)
Comment (tekknolagi) [reply zjit/src/hir/opt_tests.rs:22937]: oh! interesting. hm.

File: zjit/src/hir/opt_tests.rs (hunk 3 above)
Comment (tekknolagi) [reply zjit/src/hir/opt_tests.rs:22860]: sure, but i don't think we optimize this at all

Comment (nirvdrum) [reply zjit/src/hir/opt_tests.rs:22860]: Fair enough. I mentioned it because this PR adjusts those cases as well. I personally don't mind tests that show we don't currently optimize. They're good ways to scan for opportunities in our tests and make any future changes easier to see.

With that said, this PR does optimize `nil` block arg:

```ruby
class A
  def foo(&blk) = 42
end
class B
  def foo(&blk) = 43
end
def test(obj, &blk)
  obj.foo(&blk)
end

30.times { test(A.new); test(B.new) }
```

Run with `./build/miniruby --zjit-dump-hir-opt --zjit-call-threshold=3 test.rb`, you'll see optimized HIR not present in _master_, because the receiver was refined which lets us get to the `nil` block arg elimination.

Comment (tekknolagi) [reply zjit/src/hir/opt_tests.rs:23271]: added test in 5cfa645dff4232be38b2860e3b74a7aee2b7c69f

File: zjit/src/hir/opt_tests.rs (hunk 2 above)
Comment (tekknolagi) [reply zjit/src/hir/opt_tests.rs]: addressed in 0bbf9d75c75e5d0f9d4ac2c179c8b131d98d1171

File: zjit/src/hir.rs [hunk 4]
```diff
@@ -10268,8 +10274,55 @@ fn add_iseq_to_hir(
                         None
                     };
                     let caller_splat_length = fun.monomorphic_caller_splat_length(call_info, exit_id);
-                    let send = fun.push_insn(block, Insn::Send { recv, cd, block: block_handler, args, caller_splat_length, state: exit_id, reason: Uncategorized(opcode.into()) });
-                    state.stack_push(send);
+                    if let Some(summary) = fun.polymorphic_summary(&profiles, recv, exit_id) {
```
Comment (nirvdrum) [inline zjit/src/hir.rs]: I'm guessing you copy & pasted this from hir.rs:10200-10248. The difference between the two is only in what we set the `block` property in the `Send` instruction construction. Do you expect these will deviate? Should we extract the shared logic to a local helper function?

Comment (nirvdrum) [inline zjit/src/hir.rs:10277]: I'm guessing you copy & pasted this from hir.rs:10200-10248. The difference between the two is only in what we set the block property in the Send instruction construction. Do you expect these will deviate? Should we extract the shared logic to a local helper function?

Comment (tekknolagi) [reply zjit/src/hir.rs:10277]: we likely should and Kokubun alluded to sharing more in the future. there are a couple "send-like" handlers we can probably share
